### PR TITLE
Jb43884

### PR DESCRIPTION
**Pull request by jlehtoranta**
Add support for proxying the currently active mpris player interface to bluez5

### CODE REVIEW

File: src/mprisbluezproxy.cpp [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+
+MprisBluezProxy::MprisBluezProxy(QObject *parent)
+    : QObject(parent)
+    , m_currentController(NULL)
+    , m_bluezRegistered(false)
+{
+    qmlInfo(this) << "Creating MprisBluezProxy";
+
+    // Create mpris bluez player and make connections for it
+    m_bluezPlayer = QSharedPointer<MprisPlayer>(new MprisPlayer(this));
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:57]: Is there reason to have this as QSharedPointer when it's also parented? Or actually, is there reason to have it even dynamically allocated?

File: src/mprisbluezproxy.cpp [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+
+static const QString mprisObjectPath = QStringLiteral("/org/mpris/MediaPlayer2");
+static const QString dBusBluezService = QStringLiteral("org.bluez");
+static const QString dBusBluezMediaInterface = QStringLiteral("org.bluez.Media1");
+static const QString dBusBluezMediaRegisterMethod = QStringLiteral("RegisterPlayer");
+static const QString dBusBluezMediaUnregisterMethod = QStringLiteral("UnregisterPlayer");
+
+MprisBluezProxy::MprisBluezProxy(QObject *parent)
+    : QObject(parent)
+    , m_currentController(NULL)
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:51]: nullptr perhaps nowadays, NULL not that common in Qt code.

File: src/mprisbluezproxy.cpp [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+
+    QDBusConnection connection = QDBusConnection::systemBus();
+    if (!connection.isConnected()) {
+        qmlInfo(this) << "Failed attempting to connect to DBus system bus";
+        return;
+    }
+
+    m_bluezWatcher = new QDBusServiceWatcher(dBusBluezService,
+                                            connection,
+                                            QDBusServiceWatcher::WatchForRegistration \
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:88]: Unnecessary \

File: src/mprisbluezproxy.cpp [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+                                            | QDBusServiceWatcher::WatchForUnregistration,
+                                            this);
+    connect(m_bluezWatcher, SIGNAL(serviceRegistered(QString)), this, SLOT(onBluezServiceRegistered(QString)));
+    connect(m_bluezWatcher, SIGNAL(serviceUnregistered(QString)), this, SLOT(onBluezServiceUnregistered(QString)));
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:92]: New style signal/slot connection also here?

File: src/mprisbluezproxy.cpp [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+}
+
+QVariantMap MprisBluezProxy::generateEmptyMetadata()
+{
+    QVariantMap metadata;
+
+    metadata.insert(Mpris::enumerationToString<Mpris::Metadata>(Mpris::Title), QStringLiteral(""));
+    metadata.insert(Mpris::enumerationToString<Mpris::Metadata>(Mpris::Artist), QStringLiteral(""));
+    metadata.insert(Mpris::enumerationToString<Mpris::Metadata>(Mpris::Album), QStringLiteral(""));
+    metadata.insert(Mpris::enumerationToString<Mpris::Metadata>(Mpris::Genre), QStringLiteral(""));
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:178]: Could use just QString() on these?

File: src/mprisbluezproxy.cpp [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+    }
+}
+
+// Mpris2 Root Interface (Controller)
+void MprisBluezProxy::identityChanged() {
+    m_bluezPlayer->setIdentity(m_currentController->identity());
+}
+
+// Mpris2 Player Interface (Controller)
+void MprisBluezProxy::canControlChanged() {
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:283]: Nitpick: couple inconsistent opening braces here.

File: src/mprisbluezproxy.cpp [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+        disconnect(m_currentController.data(), &MprisController::maximumRateChanged, this, &MprisBluezProxy::maximumRateChanged);
+        disconnect(m_currentController.data(), &MprisController::metadataChanged, this, &MprisBluezProxy::metadataChanged);
+        disconnect(m_currentController.data(), &MprisController::minimumRateChanged, this, &MprisBluezProxy::minimumRateChanged);
+        disconnect(m_currentController.data(), &MprisController::playbackStatusChanged, this, &MprisBluezProxy::playbackStatusChanged);
+        disconnect(m_currentController.data(), &MprisController::positionChanged, this, &MprisBluezProxy::positionChanged);
+        disconnect(m_currentController.data(), &MprisController::rateChanged, this, &MprisBluezProxy::rateChanged);
+        disconnect(m_currentController.data(), &MprisController::shuffleChanged, this, &MprisBluezProxy::shuffleChanged);
+        disconnect(m_currentController.data(), &MprisController::volumeChanged, this, &MprisBluezProxy::volumeChanged);
+        disconnect(m_currentController.data(), &MprisController::seeked, this, &MprisBluezProxy::seeked);
```
Comment (pvuorela) [inline src/mprisbluezproxy.cpp:494]: Unless we care about external connections, could just do disconnect(controller, 0, this, 0)?

File: src/src.pro [hunk 8]
```diff
@@ -34,19 +35,22 @@ HEADERS += \
...
+    mprismanager.h \
+    mprisbluezproxy.h
 
 INSTALL_HEADERS = \
     MprisQt \
     Mpris \
     MprisPlayer \
     MprisController \
     MprisManager \
+    MprisBluezProxy \
```
Comment (pvuorela) [inline src/src.pro:47]: Does this need to be exposed as public class?

File: src/mprisbluezproxy.h [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+Q_DECLARE_METATYPE(InterfaceList)
+Q_DECLARE_METATYPE(AdapterList)
+Q_DECLARE_METATYPE(ManagedObjectList)
+
+class MprisPlayer;
+class MPRIS_QT_EXPORT MprisBluezProxy : public QObject
+{
+    Q_OBJECT
+    Q_PROPERTY(bool useSystemBus READ useSystemBus)
```
Comment (pvuorela) [inline src/mprisbluezproxy.h:61]: Doesn't seem like useful property when it's always true.

File: src/mprisbluezproxy.h [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+public:
+    MprisBluezProxy(QObject* parent = 0);
+    ~MprisBluezProxy();
+
+public slots:
+    bool useSystemBus() const {
+        return true;
+    }
```
Comment (pvuorela) [inline src/mprisbluezproxy.h:70]: In addition to above, this is now both property and slot. Doubt it needs either.

File: src/mprisbluezproxy.h [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    void metadataChanged();
+    void minimumRateChanged();
+    void playbackStatusChanged();
+    void positionChanged(qlonglong position);
+    void rateChanged();
+    void shuffleChanged();
+    void volumeChanged();
+    void seeked(qlonglong position);
+
+private Q_SLOTS:
```
Comment (pvuorela) [inline src/mprisbluezproxy.h:110]: As minor remark this mixes slots and Q_SLOTS.

File: src/mprismanager.cpp [hunk 12]
```diff
@@ -154,6 +163,35 @@ void MprisManager::setSingleService(bool single)
...
+{
+    if (m_mprisBluezProxyEnabled == enabled) {
+        return;
+    }
+
+    if (enabled) {
+        // Create MprisBluezProxy instance
+        qmlInfo(this) << "Enabling MprisBluezProxy";
+        if (m_mprisBluezProxy.isNull()) {
+            m_mprisBluezProxy = QSharedPointer<MprisBluezProxy>(new MprisBluezProxy(this));
```
Comment (pvuorela) [inline src/mprismanager.cpp:181]: Not sure if this either needs QSharedPointer when it's already parented.
For small thing the m_mprisBluezProxyEnabled member is also redundant, the same as pointer not null.

File: src/mprismanager.cpp [hunk 13]
```diff
@@ -154,6 +163,35 @@ void MprisManager::setSingleService(bool single)
...
+        }
+    } else {
+        // Delete MprisBluezProxy instance
+        qmlInfo(this) << "Disabling MprisBluezProxy";
+        if (!m_mprisBluezProxy.isNull()) {
```
Comment (pvuorela) [inline src/mprismanager.cpp:186]: Minor, but unlikely benefits much from non-null check.

File: src/mprisplayer.cpp [hunk 14]
```diff
@@ -66,7 +67,17 @@ MprisPlayer::MprisPlayer(QObject *parent)
     , m_shuffle(false)
     , m_volume(0)
 {
+    if (parent != NULL) {
+        if (parent->property("useSystemBus").isValid()) {
+            qmlInfo(this) << "MprisPlayer using system bus";
+            m_dbusType = QDBusConnection::SystemBus;
```
Comment (pvuorela) [inline src/mprisplayer.cpp:73]: Commented elsewhere already, but this feels like a thing that should be passed on the ctor parameters.

**Comment by mkosola** [comment]
Git tree moved to https://git.merproject.org/mer-core/qtmpris

**Comment by mkosola** [comment]
Git tree moved to https://git.merproject.org/mer-core/qtmpris

**Pull request closed without merging**